rootfs-builder/rootfs.sh

```diff
@@ -0,0 +1,150 @@
+if [ -n "$DEBUG" ] ; then
+	set -x
+fi
+
+#$1: Error code if want to exit differnt to 0
+usage(){
+	error="${1:-0}"
+	cat <<EOT
+USAGE: Build a Guest OS rootfs for Kata Containers image
+${script_name} [options] <distro_name>
+
+<distro_name> : Linux distribution to use as base OS.
+
+Supported Linux distributions:
+
+$(get_distros)
+
+Options:
+-a  : agent version DEFAULT: ${AGENT_VERSION} ENV: AGENT_VERSION 
+-h  : Show this help message
+-r  : rootfs directory DEFAULT: ${ROOTFS_DIR} ENV: ROOTFS_DIR
+
+ENV VARIABLES:
+GO_AGENT_PKG: Change the golang package url to get the agent source code
+            DEFAULT: ${GO_AGENT_PKG}
+AGENT_BIN   : Name of the agent binary (needed to check if agent is installed)
+USE_DOCKER: If set will build rootfs in a Docker Container (requries docker)
+            DEFAULT: not set
+EOT
+exit "${error}"
+}
+
+die()
+{
+	msg="$*"
+	echo "ERROR: ${msg}" >&2
+	exit 1
+}
+
+info()
+{
+	msg="$*"
+	echo "INFO: ${msg}" >&2
+}
+
+OK()
+{
+	msg="$*"
+	echo "INFO: [OK] ${msg}" >&2
+}
+
+get_distros() {
+	cdirs=$(find "${script_dir}" -maxdepth 1 -type d)
+	find ${cdirs} -maxdepth 1 -name "${LIB_SH}" -printf '%H\n' | while read dir; do
+		basename "${dir}"
+	done
+}
+
+
+check_function_exist() {
+	function_name="$1"
+	[ "$(type -t ${function_name})" == "function" ] || die "${function_name} function was not defined"
+}
+
+
+while getopts c:hr: opt
+do
+	case $opt in
+		a)	AGENT_VERSION="${OPTARG}" ;;
+		h)	usage ;;
+		r)	ROOTFS_DIR="${OPTARG}" ;;
+	esac
+done
+
+shift $(($OPTIND - 1))
+
+distro="$1"
+
+[ -n "${distro}" ] || usage 1
+distro_config_dir="${script_dir}/${distro}"
+
+[ -d "${distro_config_dir}" ] || die "Not found configuration directory ${distro_config_dir}"
+rootfs_lib="${distro_config_dir}/${LIB_SH}"
+source "${rootfs_lib}"
+rootfs_config="${distro_config_dir}/config.sh"
+source "${rootfs_config}"
+
+check_function_exist "build_rootfs"
+
+if [ -n "${USE_DOCKER}" ] ; then
+	image_name="${distro}-rootfs-osbuilder"
+
+	docker build  \
+		--build-arg http_proxy="${http_proxy}" \
+		--build-arg https_proxy="${https_proxy}" \
+		-t "${image_name}" "${distro_config_dir}"
+
+	docker run  \
+		--runtime runc  \
+		--env https_proxy="${https_proxy}" \
+		--env http_proxy="${http_proxy}" \
+		--env AGENT_VERSION="${AGENT_VERSION}" \
+		--env ROOTFS_DIR="/rootfs" \
+		--env GO_AGENT_PKG="${GO_AGENT_PKG}" \
+		--env AGENT_BIN="${AGENT_BIN}" \
+		--env GOPATH="${GOPATH}" \
+		-v "${script_dir}":"/osbuilder" \
+		-v "${ROOTFS_DIR}":"/rootfs" \
+		-v "${GOPATH}":"${GOPATH}" \
+		${image_name} \
+		bash /osbuilder/rootfs.sh "${distro}"
+
+	exit $?
+fi
+
+mkdir -p ${ROOTFS_DIR}
+build_rootfs ${ROOTFS_DIR}
+
+info "Check init is installed"
+[ -x "${ROOTFS_DIR}/sbin/init" ] || die "/sbin/init is not installed int ${ROOTFS_DIR}"
+OK "init is installed"
+
+info "Pull Agent source code"
+go get -d "${GO_AGENT_PKG}" || true
+OK "Pull Agent source code"
+
+info "Build agent"
+pushd "${GOPATH}/src/${GO_AGENT_PKG}"
+make HAVE_SYSTEMD=yes
```

Comment:
Doesn't this depend on whether `PACKAGES` contains `systemd`? If we need that, we should check and error if it doesn't.

Reply:
That is a good question. Have we ever tried a systemd free image. I do not think we ever did it. However @bergwolf @gnawux should be able to provide us a configuration for a non systemd rootfs.

Reply:
Also this will change the kernel command line we present when launching the VM. So we should be able to flag a systemd vs non systemd image easily.

Reply:
I am adding this here to build the agent and install systemd service files (if I build a docker container it does not has systemd).

I've never run an image without systemd, but note that I check if `/sbin/init` is installed so no matter what is the init system if it properly start the agent to communicate with the `runtime/shim/proxy` 

would be really helpful if you have an configuration to build a rootfs without systemd but other init system.

Reply:
@gnawux @bergwolf  ^^